File: src/Python/1-100/47.PermutationsII.py
from copy import deepcopy

class Solution:
    def permuteUnique(self, nums: 'List[int]') -> 'List[List[int]]':
        ans = []
        tmp = []
        m = [0] * len(nums)
        def recursive(i, nums, ans, tmp, m):
            s = set()
            if len(tmp) == len(nums):
                return ans.append(deepcopy(tmp))
            for j in range(len(nums)):
                if m[j] == 0 and nums[j] not in s:
                    s.add(nums[j])
                    tmp.append(nums[j])
                    m[j] = 1
                    recursive(j + 1, nums, ans, tmp, m)
                    m[j] = 0
                    tmp.pop()
        recursive(0, nums, ans, tmp, m)
        return ans

if __name__ == "__main__":
    solution = Solution()
    print(solution.permuteUnique([1, 1, 2]))
    print(solution.permuteUnique([1, 2, 3]))
    print(solution.permuteUnique([2, 2]))

   

File: src/CSharp/1-100/47.PermutationsII.cs
using System;
using System.Text;
using System.Collections.Generic;


namespace PeefyLeetCode.PermutationsII
{
public class Solution {
    void dfs(IList<IList<int>> ans, int[] visited, IList<int> path, int[] nums){
        HashSet<int> hash = new HashSet<int>();
        if (path.Count == nums.Length)
            ans.Add(new List<int>(path));
        else{
            for (int i = 0;i < nums.Length;++i){
                if (visited[i] == 0 && hash.Contains(nums[i]) == false){
                    hash.Add(nums[i]);
                    visited[i] = 1;
                    path.Add(nums[i]);
                    dfs(ans, visited, path, nums);
                    path.RemoveAt(path.Count - 1);
                    visited[i] = 0;
                }
            }
        }
    }
    public IList<IList<int>> PermuteUnique(int[] nums) {
        IList<IList<int>> ans = new List<IList<int>>();
        int[] visited = new int[nums.Length];      
        IList<int> path = new List<int>();
        dfs(ans, visited, path, nums);
        return ans;
    }
}
}